subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2092 @@
+	}
+
+	return error;
+}
+
+static void bt_hfp_ag_set_in_band_ring(struct bt_hfp_ag *ag, void *user_data)
+{
+	if (0 != (ag->ag_features & BT_HFP_AG_FEATURE_INBAND_RINGTONE)) {
+		int error = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BSIR:1\r\n");
+
+		if (error == 0) {
+			atomic_set_bit(ag->flags, BT_HFP_AG_INBAND_RING);
+		} else {
+			atomic_clear_bit(ag->flags, BT_HFP_AG_INBAND_RING);
+		}
+	}
+}
+
+static int bt_hfp_ag_cmer_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err)
+{
+	uint32_t number;
+	int error;
+	static const uint32_t command_line_prefix[] = {3, 0, 0};
+	bool ret;
+
+	*err = CME_ERROR_AG_FAILURE;
+
+	ret = is_char(buf, '=');
+	if (ret != true) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(command_line_prefix); i++) {
+		error = get_number(buf, &number);
+
+		if (error != 0) {
+			return error;
+		}
+
+		ret = is_char(buf, ',');
+		if (ret != true) {
+			return -EINVAL;
+		}
+
+		if (command_line_prefix[i] != number) {
+			return -EINVAL;
+		}
+	}
+
+	error = get_number(buf, &number);
+	if (error != 0) {
+		return error;
+	}
+
+	ret = is_char(buf, '\r');
+	if (ret != true) {
+		return -EINVAL;
+	}
+
+	if (number == 1) {
+		atomic_set_bit(ag->flags, BT_HFP_AG_CMER_ENABLE);
+		bt_hfp_ag_set_state(ag, BT_HFP_CONNECTED);
+		error = hfp_ag_next_step(ag, bt_hfp_ag_set_in_band_ring, NULL);
+		return error;
+	} else if (number == 0) {
+		atomic_clear_bit(ag->flags, BT_HFP_AG_CMER_ENABLE);
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int bt_hfp_ag_chld_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err)
+{
+	return -EINVAL;
+}
+
+static int bt_hfp_ag_bind_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *err)
+{
+	uint32_t number;
+	uint32_t hf_indicators = 0U;
+	int error;
+	char *data;
+	uint32_t len;
+	bool ret;
+
+	if (!((ag->ag_features & BT_HFP_AG_FEATURE_HF_IND) &&
+	      (ag->hf_features & BT_HFP_HF_FEATURE_HF_IND))) {
+		*err = CME_ERROR_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	*err = CME_ERROR_AG_FAILURE;
+
+	ret = is_char(buf, '?');
+	if (ret == true) {
+		ret = is_char(buf, '\r');
+		if (ret != true) {
+			return -EINVAL;
+		}
+
+		hf_indicators = ag->hf_indicators_of_hf & ag->hf_indicators_of_ag;
+		len = (sizeof(hf_indicators) * 8) > HFP_HF_IND_MAX ? HFP_HF_IND_MAX
+								   : (sizeof(hf_indicators) * 8);
+		for (int i = 1; i < len; i++) {
+			if (BIT(i) & hf_indicators) {
+				error = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BIND:%d,%d\r\n", i,
+							 1);
+			} else {
+				error = hfp_ag_send_data(ag, NULL, NULL, "\r\n+BIND:%d,%d\r\n", i,
+							 0);
+			}
+			if (error != 0) {
+				return -EINVAL;
+			}
+			if (hf_indicators == 0) {
+				break;
+			}
+		}
+		return 0;
+	}
+
+	ret = is_char(buf, '=');
+	if (ret != true) {
+		return -EINVAL;
+	}
+
+	ret = is_char(buf, '?');
+	if (ret == true) {
+		ret = is_char(buf, '\r');
+		if (ret != true) {
+			return -EINVAL;
+		}
+
+		data = &ag->buffer[0];
```

Comment:
Should you be checking that there is enough room in the buffer for everything you're going to encode into it?

Reply:
Yes, The buffer overflow is avoided [here ](https://github.com/zephyrproject-rtos/zephyr/pull/70532/files/7e026960d362bb5a4965f1695a48340fa55d6f0c#diff-1fc792fcea66bb7f9cec63992549978997e1eb9d77148f04092724adde5feeb2R589-R591).

`				int length = snprintk(
					data, (char *)&ag->buffer[HF_MAX_BUF_LEN - 1] - data - 3,
					"%d", i);`